### DISCUSSION TITLE
Is it possible to execute a literal/raw sql statement?

### DISCUSSION THREAD

**Discussion by DanCardin** [sqlalchemy/sqlalchemy#9802] [Usage Questions] [answered]
Essentially, I have an unknown sql string that happens to contain some substring like `'... :foo ...'`, and the `conn.execute(text(...this query))` call produces:

```
(sqlalchemy.exc.InvalidRequestError) A value is required for bind parameter 'foo'
```

Essentially I want the `execute` call to forgo any attempt to bind the given query string/definition. I thought maybe `literal` would do this:

```
conn.execute(
    text(f"CREATE VIEW some_name AS :view_def"),
    {"view_def": literal(definition)},
)
```

but that produces the same error.
---
The longer story is that in order to normalize a view definition's format, it's seemingly necessary to route some known view definition from `pg_views.definition` back through a `CREATE VIEW` statement in order to ensure you get back the same definition text.

And should that original view definition text contain the aforementioned parameter-like substring, then it's not possible (that i can see) to escape it once it's some unknown sql string blob.

**Comment by gordthompson** [comment]
You could try using

https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Connection.exec_driver_sql

**Reply by DanCardin** [reply]
It seems that that would solve for this particular case where my string contains sqlalchemy's parameter syntax of `:asdf`, but if my driver is psycopg2, and the same thing happens with `%s` or `%(id)s`, then this fails to the same sort of problem.

I want like...an explicitly non-parametrized query.

**Reply by DanCardin** [reply]
although
```python
conn.execution_options(no_parameters=True).exec_driver_sql(command)
```
maybe **does** appear to work...at least with psycopg2 from a brief test

**Reply by CaselIT** [reply]
> It seems that that would solve for this particular case where my string contains sqlalchemy's parameter syntax of `:asdf`, but if my driver is psycopg2, and the same thing happens with `%s` or `%(id)s`, then this fails to the same sort of problem.
> 
> I want like...an explicitly non-parametrized query.

These are out of the control of sqlalchemy. `exec_driver_sql` as the name implies passes the string to the driver directly, so the same rules as when using the driver directly apply. In general for psycopg `%` need to be doubled

**Comment by zzzeek** [comment, marked as answer]
if you truly want to talk to the driver and exec_driver_sql() is not good enough, get yourself a [cursor](https://docs.sqlalchemy.org/en/20/core/connections.html#working-with-the-dbapi-cursor-directly) and there you go.

that said, I dont see what the problem with text() is, colon signs get [backslash escaped](https://docs.sqlalchemy.org/en/20/core/sqlelement.html#sqlalchemy.sql.expression.text), and that's it.  it's fully consistent on any driver:

```py
conn.execute(
    text(f"CREATE VIEW some_name AS {definition.replace(':', '\:')}"),
)
```

**Reply by DanCardin** [reply]
I was concerned that perhaps that in some arbitrary sql (casts off the top of my head `select 'foo'::text`) that escaping colons could have averse side-effects?

If i'm wrong and there's some preprocessing that occurs before it hits the driver that would undo spurious `\:`s that **arent** parameters, then this would seem to be the way to go.

**Reply by gordthompson** [reply]
Yes, `text()` takes care of that:

```python
import sqlalchemy as sa

engine = sa.create_engine("postgresql://scott:tiger@192.168.0.199/test")

sql = "select 'foo'::text"
qry = sa.text(sql.replace(":", r"\:"))
engine.echo = True
with engine.begin() as conn:
    result = conn.execute(qry).scalar()
"""
2023-05-18 07:52:00,368 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2023-05-18 07:52:00,369 INFO sqlalchemy.engine.Engine select 'foo'::text
2023-05-18 07:52:00,369 INFO sqlalchemy.engine.Engine [generated in 0.00039s] {}
2023-05-18 07:52:00,389 INFO sqlalchemy.engine.Engine COMMIT
"""
```
